### PR TITLE
Separate communication nodes

### LINKED ISSUE #176: Separate robot_communication node into a radio node and grsim node
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
The current `robot_communication` node should be separated into 2 separate nodes: 1 for radio communication, and one for grSim communication.

This will keep each type of communication separate, and allow for more specific behavior/interfaces in each.

**Describe the solution you'd like**
Separate the code that communicates with grSim into its own node, and separate the code that will deal with the radio and robot communication into another separate node.

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description
*Give a high-level description of the changes in this PR*
Split the `robot_communication` node into separate `grsim_communication` nodes and `radio_communication` nodes

### Testing Done
*Outline any testing that was done for these changes. This could be unit tests, integration tests, etc.*
N/A
### Resolved Issues
*Link any issues that this PR resolved. Eg `resolves #1, #2, and #5` (note that they MUST be specified like this so Github can automatically close them then this PR merges)*
resolves #176 

### Length Justification

N/A

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
This is going to need to be updated once #9 is merged, but I figured I'd get it up for review now.

**Comment by MathewMacDougall** [comment]
Branch updated, ready for review.

File: src/thunderbots/software/grsim_communication/grsim_backend.h
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/grsim_backend.h:31]: If this class no longer inherits from `Backend` (which I still think it should?) then we should give it a Javadoc comment.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking pretty good, main question is why did we nuke `Backend`?

File: src/thunderbots/software/grsim_communication/grsim_backend.h
Comment (MathewMacDougall) [reply src/thunderbots/software/grsim_communication/grsim_backend.h:13]: I removed the generic `Backend` interface because I think the `grsim` and `radio` nodes are going to be different enough it doesn't make sense. Particularly in the case of the `sendPrimitives` function.

For example, the `grsim` node's `sendPrimitives` function will also need to accept the current state of the Friendly Team, because that information is required to translate coordinates and simulate the motion of the robots, since that takes place in the node.

On the other hand, the `radio` node does not need this information. Because the control is performed on the robot, the `sendPrimitives` function only needs the list of `Primitives` to send.

Of course each node will also have its own specific functions. For example, the `radio` node will likely have a separate function that just sends the "vision threading" information, which is the robot and ball positions on the field. This is not required for the `grsim` node. This doesn't affect the inheritance from the generic `Backend`, but I figured it was worth mentioning for completeness.

@henrymliu to correct me if I misspoke about the radio interface.

Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/grsim_backend.h:13]: Great explanation, thank you! This seems reasonable, particularly because the interface is now enforced by ROS message type now.

**Comment by garethellis0** [comment]
Thanks for the excellent explanation above! Just one unresolved comment left.

File: src/thunderbots/software/grsim_communication/grsim_backend.h
Comment (MathewMacDougall) [reply src/thunderbots/software/grsim_communication/grsim_backend.h:31]: done

**Comment by MathewMacDougall** [comment]
PR updated. I believe @henrymliu wants to merge some radio stuff before this, so I'll hold off merging until that's done.

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
Note to self: Update the wiki with new node names. Eg. the "Running the code" section

**Review by henryymliu — APPROVED** [review]
LGTM :+1:
@MathewMacDougall I might have miscommunicated my merging situation, I wanted to merge master to my fork before this goes in. Now that's done, this can be merged to master.

**Comment by MathewMacDougall** [comment]
Gotcha, thanks for clarifying. That makes my other stuff much easier.

**Pull request merged**
